api/src/Rbac/Access/Gate.php

```diff
@@ -0,0 +1,237 @@
+    {
+        $this->storage = $storage;
+
+        $this->userResolver = function () use ($userResolver) {
+
+            $user = call_user_func($userResolver);
+            if (! $user) {
+                throw new UserNotFoundException;
+            }
+
+            return $user;
+        };
+
+        $this->defaultRoles = $defaultRoles;
+    }
+
+    /**
+     * Determine if a given ability has been defined.
+     *
+     * @param  string $ability
+     * @return bool
+     */
+    public function has($ability)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Define a new ability.
+     *
+     * @param  string $ability
+     * @param  callable|string $callback
+     * @return GateContract
+     */
+    public function define($ability, $callback)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Define a policy class for a given class type.
+     *
+     * @param  string $class
+     * @param  string $policy
+     * @return GateContract
+     */
+    public function policy($class, $policy)
+    {
+        throw new NotImplementedException;
+    }
+
+    /**
+     * Determine if the given ability should be granted.
+     *
+     * @param  string $itemName
+     * @param  array|mixed $arguments
+     * @return bool
+     */
+    public function check($itemName, $arguments = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return true;
+        }
+
+        try {
+            //default roles check
+            if (! empty($this->defaultRoles) && $this->checkAccessRecursive($itemName, $arguments)) {
+                return true;
+            }
+
+            $user = $this->resolveUser();
+        } catch (UserNotFoundException $e) {
+            return false;
+        }
+
+        $assignments = $this->getStorage()->getAssignments($user->getAuthIdentifier());
+
+        return $this->checkAccessRecursive($itemName, $arguments, $assignments);
+    }
+
+    /**
+     * Get Gate service for particular user.
+     *
+     * @param Authenticatable $user
+     * @return static
+     */
+    public function forUser(Authenticatable $user)
+    {
+        return new static($this->getStorage(), function () use ($user) {return $user;}, $this->defaultRoles);
+    }
+
+    /**
+     * Check permissions and roles recursively.
+     *
+     * @param $itemName
+     * @param $params
+     * @param array $assignments
+     * @param array $itemStack
+     * @return bool
+     */
+    protected function checkAccessRecursive($itemName, $params, $assignments = [], $itemStack = [])
+    {
+        if (! ($item = $this->getItem($itemName))) {
+            return false;
+        }
+
+        $itemStack[$item->name] = $item;
+
+        if (isset($assignments[$itemName]) || in_array($itemName, $this->defaultRoles)) {
+            foreach ($itemStack as $itemName => $item) {
+                if (! $this->executeRule($item, $params)) {
+                    return false;
```

Comment:
Does this mean the rules are fail early? So the object must fulfil _all_ requirements?